### PR TITLE
feat: test embedders

### PR DESCRIPTION
**Pull request by MaksymAI**
# Embedder Tests Refactoring with Enhanced Validation

This PR refactors the embedder test files to standardize testing patterns across all provider implementations and adds enhanced validation to catch silent failures from embedding APIs.

## Changes Made

1. **Created a common `conftest.py` in the embedders test directory with:**
   - Shared fixtures for query and document inputs
   - Common fixtures for error scenarios and invalid responses
   - Helper functions for workflow creation
   - Standardized assertions for success and failure cases

2. **Added embedding validation in the base components:**
   - Created `embedding_validators.py` to validate embedding outputs
   - Added validation to catch empty or null embeddings
   - Implemented consistent error messages across all providers

3. **Added test cases for silent API failures:**
   - Tests for empty embedding arrays (`[]`)
   - Tests for null embedding values (`None`) 
   - Standardized error handling across all providers

## Updated Test Cases

All embedder implementations now test the following scenarios:

### Text Embedders

| Scenario | Behavior | Error Type |
|----------|----------|------------|
| Valid input | **SUCCESS** - Returns embedding vector | N/A |
| Missing input | **FAILURE** - Required input missing | ValidationError |
| Empty input | **FAILURE** - Cannot embed empty text | ValidationError |
| Empty embedding response | **FAILURE** - Empty embedding rejected | ValidationError |
| Null embedding response | **FAILURE** - Null embedding rejected | ValidationError |
| Auth error | **FAILURE** - Authentication error | AuthenticationError |
| Rate limit error | **FAILURE** - Rate limit error | RateLimitError |
| API error | **FAILURE** - API error | APIError |
| Invalid model | **FAILURE** - Model not found | BadRequestError |
| Max tokens exceeded | **FAILURE** - Token limit exceeded | BadRequestError |

### Document Embedders

| Scenario | Behavior | Error Type |
|----------|----------|------------|
| Valid document | **SUCCESS** - Returns embedded documents | N/A |
| Missing input | **FAILURE** - Required input missing | ValidationError |
| Empty document list | **SUCCESS** - Returns empty list | N/A |
| Empty document content | **FAILURE** - Cannot embed empty text | ValidationError |
| Empty embedding response | **FAILURE** - Empty embedding rejected | ValidationError |
| Null embedding response | **FAILURE** - Null embedding rejected | ValidationError |
| Auth error | **FAILURE** - Authentication error | AuthenticationError |
| Rate limit error | **FAILURE** - Rate limit error | RateLimitError |
| API error | **FAILURE** - API error | APIError |
| Invalid model | **FAILURE** - Model not found | BadRequestError |
| Max tokens exceeded | **FAILURE** - Token limit exceeded | BadRequestError |

## Embedder Behavior Comparison

All embedders now demonstrate consistent behavior across different providers:

| Scenario | Bedrock | Cohere | Gemini | HuggingFace | Mistral | OpenAI | VertexAI | WatsonX |
|----------|---------|--------|--------|-------------|---------|--------|----------|---------|
| Valid input | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ |
| Missing input | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ |
| Empty input | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ |
| Empty document list | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ |
| Empty document content | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ |
| Empty embedding response | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ |
| Null embedding response | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ |
| Auth error | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ |
| Rate limit error | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ |
| API error | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ |
| Invalid model | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ |
| Max tokens exceeded | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ |

## Test Results

All 178 tests are passing with the refactored implementation, demonstrating that the standardized test structure successfully validates consistent behavior across all embedder providers, including the new validation tests for silent failure cases.

----

> [!IMPORTANT]
> Refactor embedder tests to standardize patterns, enhance validation, and add comprehensive error handling across all providers.
> 
>   - **Behavior**:
>     - Refactor embedder tests to standardize testing patterns across all providers.
>     - Add validation for empty or null embeddings in `base.py`.
>     - Implement consistent error messages across all providers.
>   - **Validation**:
>     - Add `validate_embedding()` and `validate_document_embeddings()` methods in `BaseEmbedder` to check for invalid embeddings.
>     - Raise `InvalidEmbeddingError` and `DocumentEmbeddingValidationError` for invalid embeddings.
>   - **Tests**:
>     - Add test cases for empty and null embeddings in `test_bedrock_embedders.py`, `test_cohere_embedders.py`, and 8 other files.
>     - Add tests for API errors like `AuthenticationError`, `RateLimitError`, and `BadRequestError`.
>     - Use `assert_embedder_success()` and `assert_embedder_failure()` for standardized assertions.
>   - **Fixtures**:
>     - Add shared fixtures in `conftest.py` for query and document inputs, error scenarios, and invalid responses.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=dynamiq-ai%2Fdynamiq&utm_source=github&utm_medium=referral)<sup> for a2f9e8bb098035cf9b03697aa6264e1bcbd6d871. You can [customize](https://app.ellipsis.dev/dynamiq-ai/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: dynamiq/utils/embedding_validators.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
+from typing import Any
+
+
+def validate_embedding(embedding: Any) -> tuple[bool, str]:
```
Comment (acoola) [inline dynamiq/utils/embedding_validators.py:4]: Such helpers are used just by embedders, so please move them to the embedders/base.py.

File: dynamiq/utils/embedding_validators.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        embedding: The embedding vector to validate
+
+    Returns:
+        Tuple containing:
+            - Boolean indicating if embedding is valid
+            - String containing error message if invalid, empty string if valid
+    """
+    try:
+        if embedding is None:
+            return False, "Embedding is None"
```
Comment (acoola) [inline dynamiq/utils/embedding_validators.py:18]: Please use `raise SomeCustomError("error msg")` in case of error instead of returning a tuple of (bool, error_msg). So after handle validation by try..except

**Review by acoola — COMMENTED** [review]
great job, left few comments

**Review by acoola — APPROVED** [review]
great job, +1
